madgraph/iolibs/export_v4.py

```diff
@@ -4976,7 +4976,8 @@ def write_vector_size(self, fsock):
         vector_size = banner_mod.ConfigFile.format_variable(vector_size, int, name='vector_size')
         vector_size = max(1, vector_size)
 
-        text = [" integer nb_page\n"," parameter (nb_page=%i)\n" % vector_size]
+        text = [" integer VECSIZE_MEMMAX\n"," parameter (VECSIZE_MEMMAX=%i)\n" % vector_size]
+        text += [" integer VECSIZE_USED\n"," parameter (VECSIZE_USED=VECSIZE_MEMMAX)\n"]
 
```

Comment:
This sounds ugly... VECSIZE_USED should not be put in constant memory (or I do not see the point of that variable).

Reply:
yes it is ugly for the moment, but this is an interim patch, just the first step to fix the includes...

in the next MR #24 the parameter becomes a DATA and is moved around nicely

and in my madgraph4gpu patches (since a long time...) the DATA is read from file when you do "madevent < input_file", see https://github.com/madgraph5/madgraph4gpu/blob/79a0b8f71f3a9e2922494493250c5fd8985706c0/epochX/cudacpp/gg_tt.mad/SubProcesses/P1_gg_ttx/driver.f#L105

```
      write(*,*) 'Enter #events in a vector loop (max=',VECSIZE_MEMMAX,',)'
      read(*,*) VECSIZE_USED
...
      write(*,'(a16,i6)') ' VECSIZE_USED = ', VECSIZE_USED
      if( VECSIZE_USED.gt.VECSIZE_MEMMAX .or. VECSIZE_USED.le.0 ) then
        write(*,*) 'ERROR! Invalid VECSIZE_USED = ', VECSIZE_USED
        STOP
      endif
```

This is a typical input file for madevent in my tests
https://github.com/madgraph5/madgraph4gpu/blob/79a0b8f71f3a9e2922494493250c5fd8985706c0/epochX/cudacpp/tmad/logs_ggtt_mad/log_ggtt_mad_d_inl0_hrd0.txt#L110
```
*** (2-none) EXECUTE CMADEVENT_CUDACPP x1 (create events.lhe) ***
--------------------
+1 ! Fortran bridge mode (CppOnly=1, FortranOnly=0, BothQuiet=-1, BothDebug=-2)
8192 ! Number of events in a single C++ or CUDA iteration (VECSIZE_USED)
8192 1 1 ! Number of events and max and min iterations
0.000001 ! Accuracy (ignored because max iterations = min iterations)
0 ! Grid Adjustment 0=none, 2=adjust (NB if = 0, ftn26 will still be used if present)
1 ! Suppress Amplitude 1=yes (i.e. use MadEvent single-diagram enhancement)
0 ! Helicity Sum/event 0=exact
1 ! Channel number (1-N) for single-diagram enhancement multi-channel (NB used even if suppress amplitude is 0!)
--------------------
```

Reply:
So in production, I would stay that we need to go back to the normal parameter setting for input_app.txt
since otherwise we will need to change a lot of python file that does handle the creation of those many files.
(in various ways).

This especially makes sense, since you want that parameter to be identical for all channel of integration.
The point of that file is to handle parameter which are channel dependent.

The parameter "Fortran bridge mode" should also be moved to the run_card.dat.

And for the C++ we should (really) test the following setup for the input_app.txt

*** (2-none) EXECUTE CMADEVENT_CUDACPP x1 (create events.lhe) ***
--------------------
+1 ! Fortran bridge mode (CppOnly=1, FortranOnly=0, BothQuiet=-1, BothDebug=-2)
8 ! Number of events in a single C++ or CUDA iteration (VECSIZE_USED)
1000 7 3 ! Number of events and max and min iterations
0.1 ! Accuracy (ignored because max iterations = min iterations)
2 ! Grid Adjustment 0=none, 2=adjust (NB if = 0, ftn26 will still be used if present)
1 ! Suppress Amplitude 1=yes (i.e. use MadEvent single-diagram enhancement)
0 ! Helicity Sum/event 0=exact
1 ! Channel number (1-N) for single-diagram enhancement multi-channel (NB used even if suppress amplitude is 0!)
--------------------

Since this is what is going to be used in practise. (we can change the "1000" to be a multiple of 8 obviously)

Reply:
Hi Olivier, Ok I see your points. I agree, in production we will need to put those much more hardcoded somewhere, eg in the run card (but then also VECSIZE_MEMMAX can be hardcoded back to lower numbers)... I will think about it.

One option otherwise for a runtime parameter is the parameter card?

Yes the fortran bridge should go in a runcard or a parameter card. It depends if you want to rebuild or not.

As for the value=8, I do routinely test 32. I think this is a safer bet than 8, with little problems. Or at least 16 to allow also float AVX512.